### PR TITLE
Remove methods under Implementors on trait pages

### PR DESCRIPTION
**Pull request by jsha**
As discussed at https://github.com/rust-lang/rust/issues/84326#issuecomment-842652412.

On a trait page, the "Implementors" section currently lists all methods of each implementor. That duplicates the method definitions on the trait itself, and is usually not very useful. So the implementors are collapsed by default. This PR changes rustdoc to just not render them at all. Any documentation specific to an implementor can be found by clicking through to the implementor's page.

This moves the "portability" info inside the `<summary>` tags so it is still visible on trait pages (as originally implemented in #79201). That also means it will be visible on struct/enum pages when methods are collapsed.

Add `#[doc(hidden)]` to all implementations of `Iterator::__iterator_get_unchecked` that didn't already have it. Otherwise, due to #86145, the structs/enums with those implementations would generate documentation for them, and that documentation would have a broken link into the Iterator page. Those links were already "broken" but not detected by the link-checker, because they pointed to one of the Implementors on the Iterator page, which happened to have the right anchor name.

This reduces the Read trait's page size from 128kB to 68kB (uncompressed) and from 12,125 bytes to 9,989 bytes (gzipped
Demo:

https://hoffman-andrews.com/rust/remove-methods-implementors/std/string/struct.String.html#trait-implementations
https://hoffman-andrews.com/rust/remove-methods-implementors/std/io/trait.Read.html#implementors

r? @GuillaumeGomez

### CODE REVIEW

**Comment by jsha** [comment]
Ready for review.

@bors r? @GuillaumeGomez

**Comment by GuillaumeGomez** [comment]
CI disagree. ;)

File: src/librustdoc/html/render/mod.rs [hunk 1]
```diff
@@ -1552,34 +1551,27 @@ fn render_impl(
             );
         }
     }
-    let toggled = !impl_items.is_empty() || !default_impl_items.is_empty();
-    let open_details = |close_tags: &mut String, is_collapsed: bool| {
+    let on_trait_page = matches!(*parent.kind, clean::ItemKind::TraitItem(_));
+    let has_impl_items = !(impl_items.is_empty() && default_impl_items.is_empty());
+    let toggled = !on_trait_page && has_impl_items;
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:1556]: Why do we want to check if we are on a trait page now? More like: why wasn't it an issue before?

**Comment by jsha** [comment]
If we're on a trait page, we aren't going to generate documentation below
the heading, so there will be nothing to toggle.

**Comment by jsha** [comment]
CI flagged a real bug - some trait pages, like [Error](https://doc.rust-lang.org/nightly/std/error/trait.Error.html#method.downcast_mut) have in addition to the common Implementors section, and "Implementations" section - for instance `impl dyn Error + 'static`. We _do_ want to keep the method documentation for those, because that's the only page such documentation will show up on. I'll work on fixing.

**Comment by jsha** [comment]
This is now rebased on top of #86128, and I've updated the demo. Turns out the refactoring from #86128 fixed the bug.

**Comment by jsha** [comment]
Actually, hold on, I see some more improvements to clarity I can make here.

**Comment by jsha** [comment]
Alright, ready again. Updated the PR description with some info about the latest changes:

> This moves the "portability" info inside the <summary> tags so it is still visible on trait pages (as originally implemented in #79201). That also means it will be visible on struct/enum pages when methods are collapsed.

> Add #[doc(hidden)] to all implementations of Iterator::__iterator_get_unchecked that didn't already have it. Otherwise, due to #86145, the structs/enums with those implementations would generate documentation for them, and that documentation would have a broken link into the Iterator page. Those links were already "broken" but not detected by the link-checker, because they pointed to one of the Implementors on the Iterator page, which happened to have the right anchor name.

**Comment by GuillaumeGomez** [comment]
The current display is broken for me:

Screenshot from 2021-06-09 16-22-00

But otherwise looks like a great improvement!

**Comment by jsha** [comment]
Alright, the fix for layout of `#[doc(cfg)]` (#79201) was more involved than I expected. This PR brought it up into the summary so it would still be displayed even when the docblock is hidden. But that brought it into the scope of `display: flex`, which wanted to put it on the same line as everything else.

To fix the `display: flex` issue, I wound up switching to a `float: right` layout, which I think is a better fit for this use case anyhow. I do want to flag that we adopted a `display: table` layout for these in 2016 (https://github.com/rust-lang/rust/pull/33431), followed by a `display: flex` layout for these in 2018 (https://github.com/rust-lang/rust/pull/56874), and now we're coming back around to using float.

Making the float work right meant putting the srclink, stability since, and anchor earlier in the DOM.

File: src/librustdoc/html/render/mod.rs [hunk 2]
```diff
@@ -1617,10 +1599,33 @@ fn render_impl(
     w.write_str(&close_tags);
 }
 
-fn render_impl_summary(
+fn render_rightside(
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/mod.rs:1602]: Please add documentation on this function. :)

**Comment by GuillaumeGomez** [comment]
Another question: why did you remove the tests (Especially `trait-impl-items-links-and-anchors.rs`)? Updating them would seem more logical no? As far as I can tell, you removed the methods from the implementors, not from the trait implementations, right?

File: src/librustdoc/html/render/mod.rs (hunk 2 above)
Comment (jsha) [reply src/librustdoc/html/render/mod.rs:1589]: Done 👍🏻

**Comment by jsha** [comment]
> Another question: why did you remove the tests (Especially trait-impl-items-links-and-anchors.rs)? Updating them would seem more logical no? As far as I can tell, you removed the methods from the implementors, not from the trait implementations, right?

Correct. I restored trait-impl-items-and-anchors, leaving the clauses that test for methods on "implementations" on struct/enum pages. I removed the clauses that test for methods on "implementors" on trait pages. I removed src/test/rustdoc/issue-19055.rs outright because the only clauses were testing for methods on "implementors."

File: src/librustdoc/html/render/mod.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:1589]: Thanks!

**Comment by jsha** [comment]
Update: This PR simplified not just "Implementors," but also "Implementations on Foreign Types." We still want to keep the full details for "Implementations on Foreign Types," because those details won't be documented anywhere else. For instance it is really important to show an associated type in an "Implementation on Foreign Types." I'll update to fix.

**Comment by jsha** [comment]
Alright, I've restored the stuff under "Implementations on Foreign Types," and I've restored the test that was testing for that stuff. I've also updated the demo listed in the PR description. Ready for another look.

**Comment by jsha** [comment]
I merged master and squashed / rebased a few commits to simplify the history somewhat. I also backed out a somewhat unrelated change: typedefs and associated types don't currently have a srclink. In my refactoring I originally concluded that was an overight and fixed it, but this wound up being a big enough change I decided not to change unrelated things along with it.

Ready for review; demo is up to date.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
I think we're good now. Thanks a lot!

@bors: r+

**Comment by steffahn** [comment]
> Any documentation specific to an implementor can be found by clicking through to the implementor's page.

I don’t quite understand this. With the way things stand now, I can’t see any way to click through to get to the right place quickly. If I’m looking at an impl like
https://doc.rust-lang.org/nightly/std/ops/trait.BitAnd.html#impl-BitAnd%3C%26%27_%20BTreeSet%3CT%3E%3E
the easiest way to see the _documentation_ of the method is **literally** by looking into the sourcecode via [src]. One click, and you see the documentation. Of course not rendered, but it’s probably enough. The only alternative is to click on `BTreeSet` and … _scroll, scroll, scroll_ … start searching for `BitAnd` there, which is nontrivial. In this case then, the easiest way then would be to use your browser’s search.

I don’t think this is ergonomic _at all_, it _desperately_ needs - at least - a proper link to the `impl` on the type’s documentation page.

By the way, either approach to finding the documentation still completely prohibits me from figuring out that this specific `BitAnd` implementation has extra documentation on the method in the first place. On `stable`, only _undocumented_ trait methods are collapsed, so you immediately see that there _is_ extra documentation. So IMO next to a link to the proper place where the impl is implemented, I also _need_ to have an indicator (on the `trait`’s page) telling me _if there are or aren’t_ additional documentation for some methods on the `struct`’s/`enum`’s page.

**Comment by bjorn3** [comment]
^ I agree.

**Pull request merged**